### PR TITLE
Fix saving artifacts

### LINKED ISSUE #299: Sometimes there are no server artifacts if i use `before_all()` hook
### Description

```lua
g.before_all(function()
    g.s = Server:new()
    g.s:start()
end)

g.test_bar = function()
    t.assert_equals(1, 2) -- always fail
end

g.test_foo = function()
   t.assert_equals(1,1) -- always success
end

g.after_all(function()
    g.s:drop()
end)
```
Each launch is different from the previous one.

Sometimes I see my server artifacts:
```
Failed tests:
-------------

1) server.test_bar
...//test/server_test.lua:396: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
	server -> /tmp/t/artifacts/server-Kuyf6vlgbHBY
...
```

Sometimes not:
```
1) server.test_bar
/home/pc/workspace/vk/luatest///test/server_test.lua:396: expected: 2, actual: 1
stack traceback:
	...
	[C]: in function 'xpcall'
```
This behavior has been appeared in the #296 PR.

### Actual behavior
Non-deterministic behavior, each launch is different from the previous one.

### Expected behavior
Stable deterministic behavior with save of server artifacts.

### PR DESCRIPTION
**Pull request by ochaplashkin**
### What's the problem?

When creating the logic for saving artifacts [^1] a global variable has been created as `current_test` in the `runner.lua` module. Luatest has collected all the tests and  runs a simple loop. At the beginning of each iteration the current test is written to the `current_test`.

[^1]: tarantool/luatest#296

Server saved the value of the current test. However at some point the current test in the global variable and in the server object were different (the server stored information about the previous test). Also server does not store information about which test it was used in.

### What is the solution?

Redesign the logic of working with artifacts: now each test knows which server is linked to it, and each server knows about the test.

For example:

```lua
foo = Server:new()
bar = Server:new()
foo:start()
bar:start()

-- only `foo` artifacts
g.test_with_foo = function()
    foo:exec(...)
end

-- no server artifacts
g.test_without_servers = function()
    ...
end
```

#### Common behavior with hooks

```lua
g.before_all(function()
    g.all = Server:new({alias = 'all'})
    g.all:start()
end)

g.before_each(function()
    g.each = Server:new({alias = 'each'})
    g.each:start()
end)

g.before_test('test_foo', function()
    g.test = Server:new({alias = 'test'})
    g.test:start()
end)


g.test_foo = function()
    g.internal = Server:new({alias = 'internal'})
    g.internal:start()
    ...
    t.assert(false)
end
```
Output

```
Failed tests:
-------------
...
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
	all -> /tmp/t/artifacts/all-lPoX7PIOFj8e
	internal -> /tmp/t/artifacts/internal-LztDfWPAf7Rw
	each -> /tmp/t/artifacts/each--xhguIzWMweU
	test -> /tmp/t/artifacts/test-Q-UQS2ExN_Fz
```

#### Advanced behavior with hooks

Now there is no confusion between tests and servers. Only the server that is "linked" to the test will always be specified.

```lua
g.before_all(function()
    g.all = Server:new({alias = 'all'})
    g.all:start()
end)

g.before_each(function()
    g.each = Server:new({alias = 'each'})
    g.each:start()
end)

g.test_foo = function()
    t.assert(false)
end

g.test_bar = function()
    t.assert(false)
end
```

Output

```
Failed tests:
-------------

1) artifacts.test_foo
...
artifacts:
	each -> /tmp/t/artifacts/each-UlTDVHm1ptO4
2) artifacts.test_bar
...
artifacts:
	each -> /tmp/t/artifacts/each-RJTeIJKszZJW
Ran 2 tests in 0.314 seconds, 0 succeeded, 2 failed
```

#### :heart: Dynamic linking upon use

New feature: when using any server method, we will link it dynamically to the test. (thanks for the help @igormunkin :heart: )

```lua
g.s = Server:new()

g.test_foo = function()
    t.assert_equals(1, 2)
end

g.test_bar = function()
    g.s:start()
    t.assert_equals(1, 2)
end
```

Output

```
Failed tests:
-------------

1) artifacts.test_foo
...
2) artifacts.test_bar
...
artifacts:
	server -> /tmp/t/artifacts/server-ZmK5Z2O4fPo2
Ran 2 tests in 0.106 seconds, 0 succeeded, 2 failed

```

#### :outbox_tray:  TAP 13 output

```
[077] replication-luatest/gh_6036_qsync_order_test.l>                 [ fail ]
[077] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/gh_6036_qsync_order.reject:
[077] Tarantool version is 3.0.0-alpha1-148-g9af36ae10
[077] TAP version 13
[077] 1..2
[077] # Started on Mon Aug 28 09:25:37 2023
[077] # Starting group: gh-6036
[077] r1 | Deprecated option replication_connect_quorum, please use bootstrap_strategy instead
[077] r2 | Deprecated option replication_connect_quorum, please use bootstrap_strategy instead
[077] r3 | Deprecated option replication_connect_quorum, please use bootstrap_strategy instead
[077] ok     1	gh-6036.test_qsync_order
[077] ok     2	gh-6036.test_promote_order
[077] not ok 2	gh-6036.test_promote_order
[077] #   Failure in after_all hook: ...ace/vk/tarantool/test-run/lib/luatest/luatest/server.lua:356: attempt to index field 'process' (a nil value)
[077] #   stack traceback:
[077] #   	...ol/test/replication-luatest/gh_6036_qsync_order_test.lua:51: in function <...ol/test/replication-luatest/gh_6036_qsync_order_test.lua:50>
[077] #   	...
[077] #   	[C]: in function 'xpcall'
[077] #   artifacts:
[077] #   r2 -> /tmp/t/077_replication-luatest/artifacts/rs-SnDxoviAgtRd/r2-1RGjJrLxsMtA
[077] #   r3 -> /tmp/t/077_replication-luatest/artifacts/rs-SnDxoviAgtRd/r3-i0EoH_86Yiys
[077] #   r1 -> /tmp/t/077_replication-luatest/artifacts/rs-SnDxoviAgtRd/r1-2z9fjndrBZa8
[077] # Ran 2 tests in 1.997 seconds, 1 succeeded, 1 errored
```
Artifacts:
```
/tmp/t/077_replication-luatest/artifacts/rs-SnDxoviAgtRd/
├── r1-2z9fjndrBZa8
│   ├── 00000000000000000000.snap
│   ├── 00000000000000000000.xlog
│   ├── 00000000000000000022.xlog
│   └── r1.log
├── r2-1RGjJrLxsMtA
│   ├── 00000000000000000003.snap
│   ├── 00000000000000000003.xlog
│   ├── 00000000000000000023.xlog
│   └── r2.log
└── r3-i0EoH_86Yiys
    ├── 00000000000000000006.snap
    ├── 00000000000000000006.xlog
    ├── 00000000000000000014.xlog
    └── r3.log

3 directories, 12 files
```

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #299

### CODE REVIEW

**Comment by ochaplashkin** [comment]
I found a corner case:

```lua
g.s = Server:new()
g.s:start()


g.test_foo = function()
    t.assert(true)
end

g.test_bar = function()
    t.assert(false)
end

g.after_all(function()
    g.s:get_vclock()
end)
```

Expected:

test_foo - success
test_bar - failed **without** server artifacts (we didn't use it in this test).

Actual:

```
Failed tests:
-------------

1) logs.test_bar
/Users/mac/workspace/vk/luatest///test/logs_test.lua:43: expected: a value evaluating to true, actual: false
stack traceback:
	...
	[C]: in function 'xpcall'
artifacts:
	server -> /tmp/t/artifacts/server-zzHLh70opUsx
Ran 2 tests in 0.001 seconds, 1 succeeded, 1 failed

=========================================================
Failed tests:

logs.test_bar
```

File: luatest/server.lua [hunk 1]
```diff
@@ -100,6 +100,30 @@ function Server:new(object, extra)
...
+
+    -- Each method of the server instance will be overridden by a new function
+    -- in which the association of the current test and server is performed first
+    -- and then the method itself.
+    -- It solves the problem when the server is not used in the test (should not
+    -- save artifacts) and when used.
+    for k, v in pairs(self) do
+        if type(v) == 'function' then
+            object[k] = function(...)
+                local t = rawget(_G, 'current_test')
```
Comment (ylobankov) [inline luatest/server.lua:112]: Doesn't the snippet below do the work?

```lua
local t = rawget(_G, 'current_test')
if t and t.value then 
    t = t.value
    if not object.tests[t.name] then
        object.tests[t.name] = t
        t.servers[object.id] = object
     end
end
return v(...)
```

File: luatest/server.lua [hunk 2]
```diff
@@ -186,14 +210,12 @@ function Server:initialize()
...
     end
 
-    if self.current_test == nil then
-        self.current_test = rawget(_G, 'current_test')
-        if self.current_test then
-            local prefix = fio.pathjoin(Server.vardir, 'artifacts', self.rs_id or '')
-            self.artifacts = fio.pathjoin(prefix, self.id)
-            self.current_test:add_server_artifacts(self.alias, self.artifacts)
-        end
+    if self.tests == nil then
```
Comment (ylobankov) [inline luatest/server.lua:213]: Why don't you use `if not foo then ... end`?

File: luatest/server.lua [hunk 3]
```diff
@@ -328,19 +350,30 @@ function Server:restart(params, opts)
     log.debug('Restarted server PID: ' .. self.process.pid)
 end
 
+-- Save server artifacts by copying the working directory
+-- Throws an error when the copying is not successful.
+function Server:save_artifacts()
```
Comment (ylobankov) [inline luatest/server.lua:355]: You expose this function as a part of the public API now. Do you really want this?

File: luatest/output/tap.lua [hunk 4]
```diff
@@ -28,8 +29,11 @@ function Output.mt:update_status(node)
     end
     if (node:is('fail') or node:is('error')) and self.verbosity >= self.class.VERBOSITY.VERBOSE then
         print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
-        if node.artifacts then
-            print(prefix .. node.artifacts:gsub('\n', '\n' .. prefix))
+        if utils.table_len(node.servers) > 0 then
+            print(prefix .. 'artifacts:')
+            for _, server in pairs(node.servers) do
+                print(prefix .. server.alias .. server.artifacts)
```
Comment (ylobankov) [inline luatest/output/tap.lua:35]: Looks like `->` is missing.

File: luatest/server.lua (hunk 1 above)
Comment (ochaplashkin) [reply luatest/server.lua:112]: Yes, it seems to work :+1:

File: luatest/server.lua (hunk 2 above)
Comment (ochaplashkin) [reply luatest/server.lua]: Fixed :rocket:

File: luatest/server.lua (hunk 3 above)
Comment (ochaplashkin) [reply luatest/server.lua:352]: Fixed :rocket:

File: luatest/output/tap.lua [hunk 5]
```diff
@@ -28,8 +29,11 @@ function Output.mt:update_status(node)
     end
     if (node:is('fail') or node:is('error')) and self.verbosity >= self.class.VERBOSITY.VERBOSE then
         print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
-        if node.artifacts then
-            print(prefix .. node.artifacts:gsub('\n', '\n' .. prefix))
+        if utils.table_len(node.servers) > 0 then
+            print(prefix .. 'artifacts:')
+            for _, server in pairs(node.servers) do
+                print(prefix .. server.alias .. '->' .. server.artifacts)
```
Comment (ylobankov) [inline luatest/output/tap.lua:35]: `->` should be surrounded by spaces, isn't it?

Comment (ochaplashkin) [reply luatest/output/tap.lua]: Thanks! I forgot about this formatting.
Fixed.

File: luatest/output/tap.lua [hunk 6]
```diff
@@ -28,8 +29,11 @@ function Output.mt:update_status(node)
     end
     if (node:is('fail') or node:is('error')) and self.verbosity >= self.class.VERBOSITY.VERBOSE then
         print(prefix .. node.trace:gsub('\n', '\n' .. prefix))
-        if node.artifacts then
-            print(prefix .. node.artifacts:gsub('\n', '\n' .. prefix))
+        if utils.table_len(node.servers) > 0 then
+            print(prefix .. 'artifacts:')
+            for _, server in pairs(node.servers) do
+                print(prefix .. server.alias .. ' -> ' .. server.artifacts)
```
Comment (ylobankov) [inline luatest/output/tap.lua:35]: At other places, you use the `format` function for formatting. It would be good to keep consistency.

**Review by ylobankov — APPROVED** [review]
